### PR TITLE
Add ability to create a sink with customizable buffer size

### PR DESCRIPTION
**Pull request by irinarenteria**
This PR implements the ability to create a stream with a set buffer. It's still a WIP. I'm suspecting the last test assertion is a false positive. so I'm putting this up so we can see how it behaves on CI.

This does not introduce any API-breaking changes.

**Update: Ready for review.**

### CODE REVIEW

File: test/unclogged/core_test.clj
Comment (irinarenteria) [inline test/unclogged/core_test.clj:510]: @sirsean Found out that inspecting a stream _is_ part of the public API, just not the way I had thought. Calling `description` on a stream produces the properties map!

File: test/unclogged/core_test.clj
Comment (derwolfe) [inline test/unclogged/core_test.clj]: I don't believe that this is a legal message. Messages that are pushed to the sink are maps and I *think* they need the body to be specified under the key `message` (seen in use at https://github.com/RackSec/unclogged/blob/master/src/unclogged/core.clj#L40).

Comment (irinarenteria) [reply test/unclogged/core_test.clj]: Oh, this is a great observation, @derwolfe. I know that based on the docs (and my own interaction with manifold streams) that strings are valid input, _however_ it does seem like unclogged has an expected message structure **for the unclogged sink**: a map with a `message` key, as you pointed out. I hadn't considered this. Going to give it a try and circle back when I'm done!

File: test/unclogged/core_test.clj
Comment (sirsean) [inline test/unclogged/core_test.clj]: Might want to do `(dec buffer-size)` here instead of `2`.

Comment (irinarenteria) [reply test/unclogged/core_test.clj]: This is neat; didn't know about `dec`! Looks like it won't be necessary because `dotimes` executes n-1: https://clojuredocs.org/clojure.core/dotimes This wasn't failing because the test wasn't running properly due to that stream error.

File: test/unclogged/core_test.clj
Comment (irinarenteria) [reply test/unclogged/core_test.clj]: @derwolfe the message being invalid turned out to be the culprit. Thank you! I updated the tests to compose the message properly.

**Comment by irinarenteria** [comment]
@derwolfe's suggestion fixed the error in the tests. This looks ready to land. I'd like to add a few clarifying comments in the tests. That can be done in a follow-up PR as well, since the tests should be able to stand alone without comments.

File: test/unclogged/core_test.clj
Comment (sirsean) [inline test/unclogged/core_test.clj]: The indentation of the code changed because of the new `testing` up above, but cljfmt ignores lines that start with a comment. As a result, these don't line up any more. We probably want them to line up.

Comment (irinarenteria) [reply test/unclogged/core_test.clj]: Good catch, @sirsean. Done!

**Comment by sirsean** [comment]
This looks good to me but I think it would be wise to get a 👍 from someone more familiar with the project than I am.

**Comment by irinarenteria** [comment]
@derwolfe @lvh tag, you're it!

**Comment by irinarenteria** [comment]
There's no rush, btw. So this can also be post-holiday weekend.

File: src/unclogged/core.clj
Comment (lvh) [inline src/unclogged/core.clj]: This is a close call for whether it's best-served with a separate function. Was the original intent to get a var you can redef for tests?

```clojure
(s/stream (conn-opts :buffer-size 0))
```

Of course, this has the advantage that the buffer size continues to be implicit and up to manifold, though it's unlikely that'll change :)

**Review by lvh — APPROVED** [review]
Great work, @irinarenteria!

This doesn't have to be in this PR, but it might be useful to document why bufsize matters. The "obvious" reason is to not block entire app when syslog is down and blocking puts start failing -- but some language around when you just want async puts versus blocking ones might be useful. (That should definitely not block this PR, because that usage documentation is entirely missing, and it's not very fair to make you pick up that slack as a condition for getting this feature merged -- though I'd be very happy to review that PR too :))

**Comment by lvh** [comment]
As a corollary to this I should probably start handing out clojars perms. I believe I am the only person with the permission to cut releases on Clojars, but I am no longer a project collaborator on GitHub.

File: src/unclogged/core.clj
Comment (irinarenteria) [reply src/unclogged/core.clj]: Yes, initially I had separated that logic into a separate function just to make sure it works as intended. But it ended up not needing unit tests and it's a private function anyway so it's implicitly tested. I don't have a strong opinion either way; I just found this more readable :)

I'm curious why is the 0 needed?

**Comment by irinarenteria** [comment]
Thank you, @lvh! I agree it'd be good to have this documented as well. I'll have a follow-up PR to flesh out the usage section and I'll make sure to cover buffer size too.

**Comment by irinarenteria** [comment]
I'll look into permissions and what collaborator policy we have now. I was under the impression you remained a collaborator, but we also had a user audit recently, so that might have something to do with it.

File: src/unclogged/core.clj
Comment (lvh) [reply src/unclogged/core.clj]: @irinarenteria The 0 acts as a default when the key is missing. I don't think manifold treats `(stream nil)` the same as `(stream)` (which has the zero default).

Comment (sirsean) [reply src/unclogged/core.clj]: That 0 is a default value.

```
=> (def x {:foo 1})
#'x
=> (x :foo)
1
=> (x :foo 2)
1
=> (x :foot 2)
2
```

But I don't know enough about `manifold.stream` to know whether `(s/stream 0)` (buffer size 0) is the same as `(s/stream)` (unbuffered).

Comment (irinarenteria) [reply src/unclogged/core.clj]: It _seems_ to be the same:

```
=> (s/stream 0)
<< stream: {:pending-puts 0, :drained? false, :buffer-size 0, :permanent? false, :type "manifold", :sink? true, :closed? false, :pending-takes 0, :buffer-capacity 0, :source? true} >>
```

Comment (irinarenteria) [reply src/unclogged/core.clj]: Looking at this again, `(s/stream (conn-opts :buffer-size 0))` doesn't affect readability enough to warrant abstracting this into a separate function. So I'll push that change soon.

**Comment by irinarenteria** [comment]
@lvh I made you a collaborator again. Also I addressed all the feedback so far. Let me know if this is good to go!

**Comment by lvh** [comment]
Fools! I wield ultimate power! To click merge.

Thanks again for your work on this Irina :)

**Review by lvh — APPROVED** [review]
Great work!

**Comment by lvh** [comment]
Even though I'm now a collaborator, GitHub won't let me merge this pull request, and I can't figure out why. The UX suggests I am not a collaborator.

Either way, @irinarenteria lemme know how you want to handle a new release; either now or after some of the other changes you've suggested :)

**Comment by lvh** [comment]
Closing and reopening didn't affect my ability to merge it, FWIW :)

**Comment by lvh** [comment]
There doesn't appear to be a designated code owner for this project, so ostensibly GitHub just doesn't really believe I'm a collaborator.

**Comment by irinarenteria** [comment]
Looks like RackSec members are owners by default if GitHub labels are to be believed. It might have been something in the review requirements. I updated them now so you can push changes to master. Let me know if that works!

**Review by lvh — APPROVED** [review]
Approve again just in case that changes anything.

**Comment by lvh** [comment]
Approving it twice fixed it; there appears to have been a race condition where the reviewing logic didn't realize I was a collaborator.

**Comment by irinarenteria** [comment]
Awesome 🎉 Thank you, @lvh!

**Pull request merged**
